File: lib/auth/slug.ts
const SLUG_MAX = 40;

export function slugifyName(name: string): string {
  const slug = name
    .trim()
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, SLUG_MAX);

  return slug || "dojang";
}

export function randomSlugSuffix(): string {
  return Math.random().toString(36).slice(2, 6);
}

File: backend/app/core/slug.py
from __future__ import annotations

import re
import secrets
import unicodedata

SLUG_MAX = 40
_NON_SLUG = re.compile(r"[^a-z0-9]+")


def random_slug_suffix() -> str:
    return secrets.token_hex(2)


def ascii_slug(name: str) -> str:
    slug = unicodedata.normalize("NFKD", name.strip().lower())
    slug = "".join(ch for ch in slug if not unicodedata.combining(ch))
    slug = _NON_SLUG.sub("-", slug).strip("-")
    return slug[:SLUG_MAX]


def build_dojang_slug(name: str, attempt: int) -> str:
    base = ascii_slug(name) or "dojang"
    if attempt == 0 and base != "dojang":
        return base
    return f"{base}-{random_slug_suffix()}"[:SLUG_MAX]
